### PR TITLE
Implement server-side apply upgrade and downgrade

### PR DESCRIPTION
**Pull request by julianvmodesto**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

##### (1) Allow client-side to server-side apply upgrade.

Ensure that a user can change management of an object from client-side apply to
server-side apply without conflicts.

We do this by reading the last-applied-configuration annotation for client-side apply
to identify if conflicts can be forced because `kubectl apply` owns them.

##### (2) Allow server-side apply to client-side downgrade.

For an object managed with client-side apply, a user may upgrade to
managing the object with server-side apply, then decide to downgrade.

We can support this downgrade by keeping the last-applied-configuration
annotation for client-side apply updated with server-side apply.

This was suggested by @apelisse. This is an alternative to conflicting entirely and failing a downgrade.

See this KEP update: https://github.com/kubernetes/enhancements/pull/1382

**Which issue(s) this PR fixes**:

Related to https://github.com/kubernetes/kubernetes/issues/80916, https://github.com/kubernetes/kubernetes/issues/90146

**Does this PR introduce a user-facing change?**:

```release-note
Support a smooth upgrade from client-side apply to server-side apply without conflicts, as well as support the corresponding downgrade.
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+	oldLastApplied, ok := annotations[corev1.LastAppliedConfigAnnotation]
+	if !ok {
+		return
+	}
+
+	delete(annotations, corev1.LastAppliedConfigAnnotation)
+	accessor.SetAnnotations(annotations)
+
+	newLastApplied, err := json.Marshal(obj)
```
Comment (julianvmodesto) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go:72]: Oh wait this doesn't work, I think I need runtime.Encode here

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater_test.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+	if annotations == nil {
+		t.Fatalf("could not get annotations")
+	}
+
+	lastApplied, ok := annotations[corev1.LastAppliedConfigAnnotation]
+	if !ok {
+		t.Fatalf("last applied annotation is empty")
+	}
+
+	expected := `
```
Comment (kwiesmueller) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater_test.go:96]: I think that newline is failing your test.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+		return nil
+	}
+
+	oldLastApplied, ok := annotations[corev1.LastAppliedConfigAnnotation]
+	if !ok {
+		return nil
+	}
+
+	delete(annotations, corev1.LastAppliedConfigAnnotation)
+	accessor.SetAnnotations(annotations)
```
Comment (apelisse) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go:74]: I don't think these two lines are needed or even good here. You should probably just update rather than delete and re-insert.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+	newLastApplied, err := runtime.Encode(unstructured.UnstructuredJSONScheme, obj)
+	if err != nil {
+		annotations[corev1.LastAppliedConfigAnnotation] = oldLastApplied
+		accessor.SetAnnotations(annotations)
```
Comment (apelisse) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go:79]: Ah, you re-inject here in case of an error, then why did you delete it? :-)

**Review by apelisse — COMMENTED** [review]
> This is an alternative to conflicting entirely and failing a downgrade.

I don't know if that's what this does. This updates the last applied annotation so that if you want to re-use CSA after SSA, you can. BUT, I don't think any of this will avoid the conflicts with CSA on your first SSA call?

I think it's a good step, but I don't think that's enough, and we need to be careful how we do it.

@lavalamp, do you have an opinion on maintaining the last-applied annotation so that someone can re-use CSA?

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go (hunk 3 above)
Comment (julianvmodesto) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go:74]: We need to remove the last-applied annotation when we serialize the object since it will be in the annotation itself. Here's the corresponding client-side apply method:

https://github.com/kubernetes/kubernetes/blob/875d3b534e3f6f118c3bf7f7e5a6b4c6ac6fe92a/staging/src/k8s.io/kubectl/pkg/util/apply.go#L71-L74

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+// Update implements Manager.
+func (f *lastAppliedUpdater) Update(liveObj, newObj runtime.Object, managed Managed, manager string) (runtime.Object, Managed, error) {
+	return f.fieldManager.Update(liveObj, newObj, managed, manager)
+}
+
+// Apply keeps the client-side apply annotation up to date with
+// server-side apply managed fields
+func (f *lastAppliedUpdater) Apply(liveObj runtime.Object, newObj runtime.Object, managed Managed, manager string, force bool) (runtime.Object, Managed, error) {
+	err := f.updateLastApplied(newObj)
```
Comment (julianvmodesto) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go:48]: Oops, I think we need to serialize `newObj` as the annotation, but we need to set the annotation on `liveObj` to make this work

**Comment by julianvmodesto** [comment]
> > This is an alternative to conflicting entirely and failing a downgrade.
> 
> I don't know if that's what this does. This updates the last applied annotation so that if you want to re-use CSA after SSA, you can. BUT, I don't think any of this will avoid the conflicts with CSA on your first SSA call?

Oh no, I was referencing the downgrade, not the upgrade. For downgrade from SSA to CSA, we could make CSA fail for an object that's being managed by SSA [that @jennybuckley described in the downgrade proposal](https://docs.google.com/document/d/1_YKsMttYZmIw2cSg-jH5PExHjA7FUyALAgHuXDDBcFY/edit#heading=h.2u66gomiwpuj) https://github.com/kubernetes/kubernetes/issues/90146: 

> CSA requests need to be rejected by the server if the object being applied to is already in the new format. This is slightly more difficult, since CSA requests look just like a StrategicMergePatch or a MergePatch (depending on what is supported). We can get around this by checking what apply format the object is in, in each of those patch types, and then failing if the patch is attempting to set the “lastApplied” annotation in an SSA format object.

This change you proposed will make it so that SSA to CSA downgrade is valid by keeping the last-applied annotation up to date.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go (hunk 5 above)
Comment (apelisse) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go:48]: Correct!

**Comment by apelisse** [comment]
OK, Right, yes I agree then! :-)

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go (hunk 3 above)
Comment (apelisse) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go:74]: Ah yes, your right! Thanks :-)

**Comment by lavalamp** [comment]
Why do we want to keep the client side annotation? How does that help the downgrade? What does the downgrade look like if we don't maintain this?

**Comment by lavalamp** [comment]
I agree we should support downgrade, but we should probably support it for all resources, and we definitely don't want to track the client side annotation for all resources.

I guess I expected downgrade to be something like:
* client sends old apply type (patch with annotation set) ... and maybe a --force-conflicts? We don't want people to do this accidentally
* server deletes managed fields

**Comment by lavalamp** [comment]
If we wanted to get super fancy, we could notice a GET from a kubectl client side apply command, and fabricate a client side annotation at that time. I don't recommend this but it's possible.

**Comment by apelisse** [comment]
> I agree we should support downgrade, but we should probably support it for all resources

What do you mean "all resources"? All objects in etcd or all resource types?
If you mean "all objects in etcd", I don't think it makes sense to create something new for objects that have never been applied. I think the goal is only to allow people who used CSA in the past to keep using it (even though they maybe gave a try, possibly by accident, to SSA).

The goal here is mostly:
1. People have used CSA to apply a specific object, it has the last-applied annotation,
2. They try to apply that object with SSA and it works
3. They try to apply again but using CSA (maybe because they have different users applying, maybe because they applied with SSA by accident). I think this should work. As long as we've kept the last-applied annotation in sync with what CSA would have done, it will work, as long as there is only one "applier" and one version. 

> server deletes managed fields

Not sure which managed fields? All of them? That'd be losing all the "history", which means they wouldn't be able to nicely go back to SSA.

> If we wanted to get super fancy, we could notice a GET from a kubectl client side apply command, and fabricate a client side annotation at that time. I don't recommend this but it's possible.

I don't think we would even know which manager to use to build the annotation from? And also we don't really have the values anymore.

**Comment by lavalamp** [comment]
> They try to apply again but using CSA (maybe because they have different users applying, maybe because they applied with SSA by accident). I think this should work. 

Why? I think it should be possible, but I don't think it should be transparent. Users should have to explicitly downgrade. The point of this whole exercise is to get them onto SSA... Users should only be going backwards if there's a bug in SSA they need to work around, IMO.

**Comment by apelisse** [comment]
Yeah, I think they will organically as they update their version of kubectl, since it's going to be the default. But I don't think all the users are going to have the same version, so it's very possible that they go back-and-forth without even realizing. I don't see any reason to not have the smoothest experience in that case.

**Comment by lavalamp** [comment]
> But I don't think all the users are going to have the same version, so it's very possible that they go back-and-forth without even realizing.

I don't think we're helping people if we make it work that way. People should know what their system is doing. Wouldn't you want to know if your coworker upgraded to a SSA version before they were supposed to?

Also people should definitely not be using the existing client side apply like that, it's not really safe for multiple appliers.

**Comment by apelisse** [comment]
> Also people should definitely not be using the existing client side apply like that, it's not really safe for multiple appliers.

Maybe it's a bad example, but whoever had time to re-deploy the mungebot would do it :-)
I wouldn't be surprised if it's still a super common workflow.

**Comment by apelisse** [comment]
> Wouldn't you want to know if your coworker upgraded to a SSA version before they were supposed to?

Not that much, I think it's going to happen by accident anyway. I'm a little more concerned about a coworker using CSA and maybe stomping some field they shouldn't. But even that I think is better than the alternatives I remember seeing.

**Comment by lavalamp** [comment]
> whoever had time to re-deploy the mungebot would do it :-)

...and you'd definitely want to know if the last person used SSA, don't you think?

**Comment by apelisse** [comment]
> ...and you'd definitely want to know if the last person used SSA, don't you think?

I guess the question is, what difference does it make?

**Comment by lavalamp** [comment]
It adds risk to do something different. Teams should generally be using tools at the same versions, there's just weird problems if you don't. It makes onboarding coworkers harder when they get random, different error messages.

**Comment by apelisse** [comment]
OK So, disregarding how it's done, what sort of experience do you think would be better?

I can think of a couple:
1. We error when going from CSA to SSA and require some sort of "force". That's sort-of what we're doing today, though today you'll get an error when you change fields owned by CSA, which means you can get a conflict two years after starting to use SSA, when you finally change that field that was still owned by CSA.
2. We error when going from SSA to CSA and you can force, but we can't provide that flag through older versions of kubectl. I think what makes the most sense is: if I force, I can then apply repeatedly with CSA without forcing, and then if I re-SSA (smoothly), I would need to force again to go back to CSA mode.

**Comment by lavalamp** [comment]
Yeah, the latter sounds close to what I'd expect--

Ideally, transitioning from CSA -> SSA would be seamless, but if it leaves
a landmine like you mention, then it'd be much less confusing to get a
force up-front.

Transitioning from SSA -> CSA should require a force, since we hope you
only want to do that in an exceptional case.

In either case, if you continue to use the tool that last did the apply,
you shouldn't have to force (unless someone else clobbered your object, of
course).

**Comment by kwiesmueller** [comment]
> We error when going from SSA to CSA and you can force, but we can't provide that flag through older versions of kubectl.

Why can't we use the already existing force flag in kubectl for this?

Also linking my previous comment from the other discussion: https://github.com/kubernetes/kubernetes/issues/90146#issuecomment-613692487

**Comment by kwiesmueller** [comment]
And I agree that people should be using the same tools/versions etc.
I'm not sure if that's going on out there though. I'm quite sure many people are mixed with my teams. And as there is no warning or anything when your kubectl doesn't match the server or the ones previously writing the object, i wouldn't expect a lot of users to care :-/

**Comment by apelisse** [comment]
>> We error when going from SSA to CSA and you can force, but we can't provide that flag through older versions of kubectl.
>
> Why can't we use the already existing force flag in kubectl for this?

The force flag in kubectl doesn't alter the request or does anything we want. I would be strongly opposed to changing the semantics of the flag by cherry-picking changes to older versions.

**Comment by kwiesmueller** [comment]
Ah, i didn't know it's not touching the request.
Then i agree.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go (hunk 5 above)
Comment (julianvmodesto) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go]: Okay, fixed. Will need an e2e or integration test for this.

Tested on a local cluster:

```
$ cat <<EOF > test.yaml
→ apiVersion: v1
→ kind: Pod
→ metadata:
→   namespace: default
→   name: pause
→ spec:
→   containers:
→   - image: k8s.gcr.io/pause:3.1
→     name: pause
→ EOF

$ ./cluster/kubectl.sh apply -f test.yaml
pod/pause created

$ ./cluster/kubectl.sh apply view-last-applied -f test.yaml
apiVersion: v1
kind: Pod
metadata:
  annotations: {}
  name: pause
  namespace: default
spec:
  containers:
  - image: k8s.gcr.io/pause:3.1
    name: pause

$ sed s/pause:3.1/pause:latest/g test.yaml -i

$ ./cluster/kubectl.sh apply -f test.yaml --server-side --force-conflicts
pod/pause serverside-applied

$ ./cluster/kubectl.sh apply view-last-applied -f test.yaml
apiVersion: v1
kind: Pod
metadata:
  name: pause
  namespace: default
spec:
  containers:
  - image: k8s.gcr.io/pause:latest
    name: pause

```

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater_test.go (hunk 2 above)
Comment (julianvmodesto) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater_test.go]: Hm I'm going to move this test somewhere else since the test field manager doesn't use the `merge`

**Comment by julianvmodesto** [comment]
Rebased in https://github.com/kubernetes/kubernetes/pull/84134 and converted the ApplyWithLastApplied (upgrader) to a Manager because we must look at last-applied (for the upgrade) before changing last-applied as in this change (for the downgrade).

**Comment by apelisse** [comment]
One of your test is failing for a good reason.

**Comment by julianvmodesto** [comment]
Fixed integration test + also added unit tests

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+// Apply keeps the client-side apply annotation up to date with
+// server-side apply managed fields
+func (f *lastAppliedUpdater) Apply(liveObj, newObj runtime.Object, managed Managed, manager string, userAgent string, force bool) (runtime.Object, Managed, error) {
+	// Sync the client-side apply annotation
+	// only if the manager is the default manager for kubectl server-side apply
+	if manager != "kubectl" {
+		return f.fieldManager.Apply(liveObj, newObj, managed, manager, userAgent, force)
+	}
+	setLastApplied(newObj)
+	return f.fieldManager.Apply(liveObj, newObj, managed, manager, userAgent, force)
```
Comment (apelisse) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go:56]: That could be simpler?
```golang
{
	if manager == "kubectl" {
	    setLastApplied(newObj)
	}
	return f.fieldManager.Apply(liveObj, newObj, managed, manager, userAgent, force)
 }
```

Comment (julianvmodesto) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go:55]: Oh nice, done

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+func (f *lastAppliedUpdater) Update(liveObj, newObj runtime.Object, managed Managed, manager string) (runtime.Object, Managed, error) {
+	return f.fieldManager.Update(liveObj, newObj, managed, manager)
+}
+
+// Apply keeps the client-side apply annotation up to date with
+// server-side apply managed fields
+func (f *lastAppliedUpdater) Apply(liveObj, newObj runtime.Object, managed Managed, manager string, userAgent string, force bool) (runtime.Object, Managed, error) {
+	// Sync the client-side apply annotation
+	// only if the manager is the default manager for kubectl server-side apply
+	if manager != "kubectl" {
```
Comment (apelisse) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go:52]: it should be `==`, not `!=`, right? You want to set the last applied annotation if it *is* kubectl?

Comment (apelisse) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go:52]: Make sure your test failed by the way, if it didn't then I'm concerned :D

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+		setLastApplied(newObj)
+	}
+	return f.fieldManager.Apply(liveObj, newObj, managed, manager, userAgent, force)
+}
+
+func setLastApplied(obj runtime.Object) {
```
Comment (apelisse) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go:58]: Why did you decide not to return an error here?

Comment (apelisse) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go:58]: Since this is in "Apply", we have more tolerance to failure.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go (hunk 7 above)
Comment (julianvmodesto) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go:52]: Gahh yes

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedmanager.go [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+	}
+	conflictSet := conflictsToSet(conflicts)
+
+	// Check if conflicts are allowed due to client-side apply,
+	// and if so, then force apply
+	allowedConflictSet, err := f.allowedConflictsFromLastApplied(liveObj)
+	if err != nil {
+		return newLiveObj, newManaged, newErr
+	}
+	if !conflictSet.Difference(allowedConflictSet).Empty() {
```
Comment (apelisse) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedmanager.go:85]: We probably discussed that already, but I forgot the outcome. Should we really complain about things that we believe shouldn't be conflict? In other words, shouldn't we return the difference as the set of conflicts here?

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedmanager.go [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+		return newLiveObj, newManaged, newErr
+	}
+	return f.fieldManager.Apply(liveObj, newObj, managed, manager, "", true)
+}
+
+func (f *lastAppliedManager) allowedConflictsFromLastApplied(liveObj runtime.Object) (*fieldpath.Set, error) {
```
Comment (apelisse) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedmanager.go:91]: I was wondering what solution would exist for people who have a non-parseable last-applied annotation. There already exists tool in kubectl to edit it, but I also want to mention that forcing apply should solve the problem. Would you be able to confirm that it would not result in this piece of code failing (I think it wouldn't because we only get here if there are conflicts) and that the last-applied annotation would be updated none-the less?

**Review by apelisse — COMMENTED** [review]
Thanks again, that's pretty awesome!

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedmanager.go (hunk 9 above)
Comment (julianvmodesto) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedmanager.go:85]: Oh wow right, we should do this and only return the difference

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go (hunk 7 above)
Comment (apelisse) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go]: That test failed, it's probably related?
https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/90187/pull-kubernetes-bazel-test/1273731844889645060

Comment (julianvmodesto) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go]: Fixed! Confirmed the tests failed as expected

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go (hunk 8 above)
Comment (julianvmodesto) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go]: Ah okay, returning an error here. Not sure if we should add any further error remediation message like you mentioned for the upgrade? e.g. ask user to remove the last-applied annotation or set it properly with kubectl subcommand

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedmanager.go (hunk 9 above)
Comment (julianvmodesto) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedmanager.go:85]: Done

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedmanager.go (hunk 10 above)
Comment (julianvmodesto) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedmanager.go:92]: Added a test for this, `TestServerSideApplyWithInvalidLastApplied`

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedmanager.go (hunk 9 above)
Comment (julianvmodesto) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedmanager.go:85]: Added a test for this

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+func (f *lastAppliedUpdater) Update(liveObj, newObj runtime.Object, managed Managed, manager string) (runtime.Object, Managed, error) {
+	return f.fieldManager.Update(liveObj, newObj, managed, manager)
+}
+
+// Apply keeps the client-side apply annotation up to date with
+// server-side apply managed fields
+func (f *lastAppliedUpdater) Apply(liveObj, newObj runtime.Object, managed Managed, manager string, userAgent string, force bool) (runtime.Object, Managed, error) {
+	// Sync the client-side apply annotation
+	// only if the manager is the default manager for kubectl server-side apply
+	if manager == ServerSideApplyFieldManager {
```
Comment (apelisse) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go:52]: Why are we using the manager rather than the userAgent here?

Comment (julianvmodesto) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go]: Oh, fixed and commented as discussed.

**Comment by apelisse** [comment]
Looks like the errors are spurious and/or timing-out.

**Review by apelisse — COMMENTED** [review]
/lgtm

**Comment by julianvmodesto** [comment]
This change implements both the upgrade and downgrade. Consolidated from https://github.com/kubernetes/kubernetes/pull/84134.

/assign @lavalamp
/retest

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedmanager.go [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+}
+
+// Apply will consider the last-applied annotation
+// for upgrading an object managed by client-side apply to server-side apply
+// without conflicts.
+func (f *lastAppliedManager) Apply(liveObj, newObj runtime.Object, managed Managed, manager string, userAgent string, force bool) (runtime.Object, Managed, error) {
+	newLiveObj, newManaged, newErr := f.fieldManager.Apply(liveObj, newObj, managed, manager, userAgent, force)
+	// Upgrade the client-side apply annotation only from kubectl server-side-apply.
+	// We use the user agent because the field manager from kubectl may be customized.
+	if userAgent != "kubectl" {
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedmanager.go:65]: If the user agent is customized, will the user really expect to do the upgrade logic? We definitely don't want more than one applier triggering the upgrade logic (right?), how is that avoided?

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater_test.go [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+	if err := f.Apply(appliedObj, "test-kubectl-ssa", "kubectl", false); err != nil {
+		t.Errorf("error applying object: %v", err)
+	}
+
+	lastApplied, err = getLastApplied(f.liveObj)
+	if err != nil {
+		t.Errorf("failed to get last applied: %v", err)
+	}
+
+	if lastApplied == originalLastApplied ||
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater_test.go:86]: I'd feel more comfortable if this were checked golden data, or then actually run through the kubectl CSA logic to ensure that it's actually compatible.

**Comment by lavalamp** [comment]
Sorry for delay in reviewing... Apart from what I commented on, looks great :)

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedmanager.go (hunk 12 above)
Comment (julianvmodesto) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedmanager.go]: Replaced with a check for the server-side apply field manager instead of the kubectl user agent, as discussed with @kubernetes/release-managers 

Users may opt-out by specifying a different field manager, which we will document.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater_test.go (hunk 13 above)
Comment (julianvmodesto) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater_test.go:86]: The integration test in test/cmd usees kubectl directly -- is this okay? Maybe we can do something better in test/integration/apiserver/apply/apply_test.go too.

Comment (lavalamp) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater_test.go:86]: Yeah that's probably good enough. As long as we would notice if it broke.

**Comment by lavalamp** [comment]
/approve

I think that made the change cleaner, too, thanks!

**Comment by julianvmodesto** [comment]
Oh I’ll probably need to squash

**Comment by apelisse** [comment]
Awesome @julianvmodesto !
/lgtm

**Comment by apelisse** [comment]
I'm confused, is `/retest` retrying things that have succeeded in the past? Every time you run `/retest` you have new tests failing? This is never going to end, will it? :-)

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+	// then continue because we have no annotation to update
+	if manager == "kubectl" && hasLastApplied(liveObj) {
+		lastAppliedValue, err := buildLastApplied(newObj)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to build last-applied annotation: %v", err)
+		}
+		err = setLastApplied(liveObj, lastAppliedValue)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to set last-applied annotation: %v", err)
+		}
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go:67]: this needs to check if the new annotation value causes https://github.com/kubernetes/kubectl/issues/712 and if so, I think it should clear the annotation rather than populating annotations that fail validation

Comment (julianvmodesto) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go:67]: Ah will send a change
